### PR TITLE
[SYCL] Allow for different types in group algorithms

### LINKED ISSUE #8341: Compilation failure on sycl::joint_exclusive_scan, sycl::joint_reduce and sycl::reduce_over_group
**Summary**
Compilation failures on `sycl::joint_exclusive_scan` with different types of `InPtr` and `OutPtr`, 
on `sycl::joint_reduce` with different types of `*Ptr` and `T`
and on `sycl::reduce_over_group` with different types of `V` and `T`

Although in SYCL2020 specification [4.17.4.6. exclusive_scan and inclusive_scan](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_exclusive_scan_and_inclusive_scan):
```
template <typename Group, typename InPtr, typename OutPtr,
          typename BinaryOperation>
OutPtr joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
                            BinaryOperation binary_op); // (1)
```
there is no requirement that `InPtr` and `OutPtr` types should be the same and
in SYCL2020 specification [4.17.4.5. reduce](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_reduce):
```
template <typename Group, typename Ptr, typename T, typename BinaryOperation>
T joint_reduce(Group g, Ptr first, Ptr last, T init,
               BinaryOperation binary_op); // (2)
template <typename Group, typename V, typename T, typename BinaryOperation>
T reduce_over_group(Group g, V x, T init, BinaryOperation binary_op); // (4)
```
there is no requirements that `*Ptr` and `T` or `V` and `T` types should be the same.

**To reproduce**
```
#include <stdint.h>

#include <sycl.hpp>

#include <variant>
#include <iostream>
#include <algorithm>

int main() {
  using AccT = sycl::accessor<int, 1>;
  using AccT2 = sycl::accessor<unsigned, 1>;
  
  int v[256] = {};
  std::fill_n(v, 255, 1);
  unsigned res[256] = {};
  
  constexpr size_t global_range_size = 256;
  constexpr size_t local_range_size = 2;

  sycl::range<1> global_range(global_range_size);
  sycl::range<1> local_range(local_range_size);

  {
  sycl::buffer<int,1> v_buf(v, global_range);
  sycl::buffer<unsigned,1> res_buf(res, global_range);

  sycl::nd_range<1> nd_range(global_range, local_range);
  sycl::queue queue;
  queue
      .submit([&](sycl::handler &cgh) {
        AccT v_acc( v_buf, cgh);
        AccT2 res_acc( res_buf, cgh);
        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
          auto group = item.get_group();
          int* v_begin = v_acc.get_pointer();
          int* v_end = v_begin + v_acc.size();
          unsigned* r_begin = res_acc.get_pointer();
          unsigned* r_end;

          auto op_plus = sycl::plus<unsigned>();
          r_end = sycl::joint_exclusive_scan( group, v_begin, v_end, r_begin, op_plus );
          unsigned input = 1;
          auto res_reduce = sycl::joint_reduce( group, v_begin, v_end, input, op_plus );
          int x = 2;
          auto res_rog = sycl::reduce_over_group( group, x, input, op_plus );
        });
  })
  .wait_and_throw();
  }
  return EXIT_SUCCESS;
}
```

**To build**
`clang++ -fsycl ./sample_joint_exclusive_scan.cpp -o sample`

**Output**
```
sample_joint_exclusive_scan.cpp:41:12: error: no matching function for call to 'joint_exclusive_scan'
          r_end = sycl::joint_exclusive_scan( group, v_begin, v_end, r_begin, op_plus );
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:792:1: note: candidate template ignored: requirement 'detail::is_native_op<int, std::plus<unsigned int>>::value' was not satisfied [with Group = group<1>, InPtr = int *, OutPtr = unsigned int *, BinaryOperation = sycl::plus<unsigned int>]
joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:738:1: note: candidate function template not viable: requires 6 arguments, but 5 were provided
joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result, T init,
^
sample_joint_exclusive_scan.cpp:43:22: error: no matching function for call to 'joint_reduce'
          auto res_reduce = sycl::joint_reduce( group, v_begin, v_end, input, op_plus );
                            ^~~~~~~~~~~~~~~~~~
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:330:1: note: candidate template ignored: requirement 'detail::is_native_op<int, std::plus<unsigned int>>::value' was not satisfied [with Group = group<1>, Ptr = int *, T = unsigned int, BinaryOperation = sycl::plus<unsigned int>]
joint_reduce(Group g, Ptr first, Ptr last, T init, BinaryOperation binary_op) {
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:302:1: note: candidate function template not viable: requires 4 arguments, but 5 were provided
joint_reduce(Group g, Ptr first, Ptr last, BinaryOperation binary_op) {
^
sample_joint_exclusive_scan.cpp:45:19: error: no matching function for call to 'reduce_over_group'
          auto res_rog = sycl::reduce_over_group( group, x, input, op_plus );
                         ^~~~~~~~~~~~~~~~~~~~~~~
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:249:1: note: candidate template ignored: requirement 'detail::is_native_op<int, std::plus<unsigned int>>::value' was not satisfied [with Group = group<1>, V = int, T = unsigned int, BinaryOperation = sycl::plus<unsigned int>]
reduce_over_group(Group g, V x, T init, BinaryOperation binary_op) {
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:272:1: note: candidate template ignored: requirement 'detail::is_vector_arithmetic<int>::value' was not satisfied [with Group = group<1>, V = int, T = unsigned int, BinaryOperation = sycl::plus<unsigned int>]
reduce_over_group(Group g, V x, T init, BinaryOperation binary_op) {
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:177:1: note: candidate function template not viable: requires 3 arguments, but 4 were provided
reduce_over_group(Group, T x, BinaryOperation binary_op) {
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:202:1: note: candidate function template not viable: requires 3 arguments, but 4 were provided
reduce_over_group(Group g, T x, BinaryOperation binary_op) {
^
/opt/sycl/bin/../include/sycl/group_algorithm.hpp:222:1: note: candidate function template not viable: requires 3 arguments, but 4 were provided
reduce_over_group(Group g, T x, BinaryOperation binary_op) {
^
3 errors generated.
```

### PR DESCRIPTION
**Pull request by jzc**
Closes #8341. The overloads for `reduce`, `exclusive_scan`, and `inclusive_scan` that the spec allowed to take differently typed input arguments should be supported now.

### CODE REVIEW

**Comment by jzc** [comment]
@intel/sycl-language-enabling-triage FYI
@intel/llvm-reviewers-runtime can this get a review?

File: sycl/include/sycl/group_algorithm.hpp [hunk 1]
```diff
@@ -264,10 +264,8 @@ std::enable_if_t<
     (is_group_v<std::decay_t<Group>> &&
      (detail::is_scalar_arithmetic<V>::value || detail::is_complex<V>::value) &&
      (detail::is_scalar_arithmetic<T>::value || detail::is_complex<T>::value) &&
-     detail::is_native_op<V, BinaryOperation>::value &&
      detail::is_native_op<T, BinaryOperation>::value &&
-     detail::is_plus_or_multiplies_if_complex<T, BinaryOperation>::value &&
-     detail::is_plus_or_multiplies_if_complex<V, BinaryOperation>::value),
+     detail::is_plus_or_multiplies_if_complex<T, BinaryOperation>::value),
```
Comment (steffenlarsen) [inline sycl/include/sycl/group_algorithm.hpp:268]: Should we add the requirement that `std::is_convertible_v<V, T>`?

File: sycl/test-e2e/GroupAlgorithm/different_types.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
+// UNSUPPORTED: hip
```
[the remark this answers is not in the document]
Comment (jzc) [reply sycl/test-e2e/GroupAlgorithm/different_types.cpp:1]: I based this test off of `reduce_sycl2020.cpp`/`exclusive_scan_sycl2020.cpp`/`inclusive_scan_sycl2020.cpp`, which all were unsupported for HIP, so I kept that. I'm unsure why those tests were unsupported for HIP though.

File: sycl/test-e2e/GroupAlgorithm/different_types.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <cstdint>
+#include <limits>
+#include <numeric>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+template <class OutT, class InputContainer, class InitT, class BinaryOperation>
+void test(queue q, const InputContainer &input, InitT init,
```
Comment (maarquitos14) [inline sycl/test-e2e/GroupAlgorithm/different_types.cpp:15]: You might want to pass by reference the queue as well.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

**Pull request merged**